### ISSUE TITLE
Contemplate switching (programming) language

### ISSUE THREAD

**Issue by danielat998** [danielat998/HanziToAnki#6]
While Java is well suited to this project in many ways, I am a little weary of introducing another Java dependency, in particular because I know these can be a pain in Windows, but for other reasons too. I think we should at least consider switching moving to another language now, as it would be much easier done at this stage than later on. I am also personally interested in playing around with the Rust language and think this might be a good oppoprtunity to do that, assuming that the standard libraries/external libraries are sufficient (In particular, I have no idea how pleasant designing GUIs is in Rust...).

**Comment by james-clark-5** [comment]
Maybe summarise Pros and Cons for candidate languages?

Java would mean easier usage with Android development, correct?

"Kotlin" seems to be a thing that's going to be available soon. Don't know much about it; Android support for it announced at Google I/O: https://www.reddit.com/r/Android/comments/6bqmwl/kotlin_on_android_now_official/.

In the end, how will the user scan an article for HanZi? Camera? Does this imply smartphone usage would be much more prevalent than PC usage? Would a user have to upload their images for PC for use with pc app?

**Comment by danielat998** [comment]
> Java would mean easier usage with Android development, correct?
Yes, this is a big pro for Java. (or Kotlin as you mentioned) I know native development is possible on Android but I've never tried it and imagine it's a little less streamlined than the normal route.

>Does this imply smartphone usage would be much more prevalent than PC usage?
This does raise a very valid question. My initial thinking was to keep this as a Desktop app, as that is what I personally consider to be most useful - I can just copy and paste text from a browser. However, the idea of scanning an article with a camera is also a valid one and my personal use case does not match other people's. Another use case has just occured to me while writing this - a Web UI, though this could be done with almost any language as the backend.

tl;dr: Java may well be the best compromise, but I would like to do a little research first.

**Comment by james-clark-5** [comment]
I think it's true that on Android/iOS there are already lots of language-learning apps that have in-built tools, so perhaps major value of this tool would be on a desktop where there's perhaps less tools/apps available to pair with foreign-language content consumption.

Using a web-UI would make a lot of sense, if content is consumed via an internet browser (very likely). However, this is probably not the best way of doing things for smartphone use-cases.

**Comment by danielat998** [comment]
>However, this is probably not the best way of doing things for smartphone use-cases.

I agree, ideally we would cover all three bases - smartphone App (just Android though, at least for now unless someone feels like rewriting the entire codebase...), desktop program and web UI. And this could be done using the the same core engine - quite likely in Java(Kotlin could be interesting to experiment with too and is completely compatible with Java).

**Comment by james-clark-5** [comment]
Would a desktop program be necessary if the user is consuming content through web?

**Comment by james-clark-5** [comment]
Perhaps Electron would be suitable tool to make a desktop suit, that functions similarly to the web UI:

From their homepage: "Build cross platform desktop appswith JavaScript, HTML, and CSS"

**Comment by danielat998** [comment]
> Perhaps Electron would be suitable tool to make a desktop suit, that functions similarly to the web UI:

Maybe, though if the entire functionality we possible using the Web UI, this might be slightly redundant, unless one wanted an offline version.

**Comment by danielat998** [comment]
> Would a desktop program be necessary if the user is consuming content through web? 

Not necessarily - get searching for domain names!

**Comment by james-clark-5** [comment]
Could be worth simplifying the target audience, or deciding what order to tackle the different platforms in etc., to keep things simple and so the core functionality of the program can be explored

**Comment by danielat998** [comment]
> Could be worth simplifying the target audience, or deciding what order to tackle the different platforms in etc., to keep things simple and so the core functionality of the program can be explored

Agreed, I think that there will be several use cases and so we should determine what these are and tackle them one by one (I would suggest desktop first just because I've already got that working-kind of). I also think we should consider splitting up into different projects, e.g. core, android, web, desktop and draw up some nice class diagrams. James, when you have time, you might want to familiarise yourself with UML/class diagrams. We don't need a formal specification but we should definitely do some rough sketches before we start working. In particular we should consider what the core contains, as my initial plan was for the core to be purely a dictionary library, with no flashcard code, as this would have a wide use of applications in other projects. So perhaps theflashcard functionality should be a separate layer that then sits on top of this dictionary library, though I'm not sure if this would be spreading ourselves too thin... Also I will need to look into creating linked git repos as this is something I have never done, but I think it is possible.

**Comment by danielat998** [comment]
I think we have pretty much settled for Java, in a large part due to Android compatibility. Kotlin could also be experimented with for some files if anyone feels like it as it interoperates with Java.
